src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/HtmlCSharp/DefaultLSPProgressListener.cs

```diff
@@ -0,0 +1,136 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.LanguageServer.Client;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Newtonsoft.Json.Linq;
+using System.Composition;
+
+namespace Microsoft.VisualStudio.LanguageServerClient.Razor.HtmlCSharp
+{
+    [Shared]
+    [Export(typeof(LSPProgressListener))]
+    internal class DefaultLSPProgressListener : LSPProgressListener, IDisposable
+    {
+        private readonly ILanguageServiceBroker2 _languageServiceBroker;
+
+        [ImportingConstructor]
+        public DefaultLSPProgressListener(ILanguageServiceBroker2 languageServiceBroker)
+        {
+            if (languageServiceBroker is null)
+            {
+                throw new ArgumentNullException(nameof(languageServiceBroker));
+            }
+
+            _languageServiceBroker = languageServiceBroker;
+
+            _languageServiceBroker.ClientNotifyAsync += ClientNotifyAsyncListenerAsync;
+        }
+
+        private ConcurrentDictionary<string, CallbackRequest> ActiveRequests { get; }
+            = new ConcurrentDictionary<string, CallbackRequest>();
+
+        public override async Task ClientNotifyAsyncListenerAsync(object sender, LanguageClientNotifyEventArgs args)
+        {
+            await ProcessProgressNotificationAsync(args.MethodName, args.ParameterToken).ConfigureAwait(false);
+        }
+
+        private async Task ProcessProgressNotificationAsync(string methodName, JToken parameterToken)
+        {
+            if (methodName != Methods.ProgressNotificationName ||
+               !parameterToken.HasValues ||
+               parameterToken["value"] is null ||
+               parameterToken["token"] is null)
+            {
+                return;
+            }
+
+            var token = parameterToken["token"].ToObject<string>(); // IProgress<object>>();
+
+            if (string.IsNullOrEmpty(token) || !ActiveRequests.TryGetValue(token, out var callbackRequest))
+            {
+                return;
+            }
+
+            var value = parameterToken["value"];
```

Comment:
in FAR's case `value` represents the serialized form of the FAR results right?

Reply:
Yes, serialized `VSReferences[]` which is deserialized in the callback
https://github.com/dotnet/aspnetcore-tooling/pull/2119/files#diff-4d367dc180f9d8b2a85502e662441dc9R158